### PR TITLE
Fix a bug involving the AutoUnloadPool and iterators (second attempt)

### PR DESCRIPTION
**Pull request by brummett**
The Context Loading Iterator buffers one object to ensure that all the joined
rows have been loaded before returning the primary object via the iterator.
This interacted badly with the AutoUnloadPool when the pool was created before
the object was pulled off of the iterator by also unloading the buffered object

This changes the AutoUnloadPool to use the same weakening mechanism that the cache pruner uses.  It fixes the bug because the Loading Iterator maintains a reference to the buffered object, keeping it from getting unloaded until it's not buffered anymore.

This PR supersedes #91

### CODE REVIEW

File: lib/UR/Object.pm [hunk 1]
```diff
@@ -740,20 +740,29 @@ sub DESTROY {
...
-            $UR::Context::all_objects_loaded->{$class}{$obj->{id}} = $obj;
+            die "Object found in all_objects_loaded does not match destroyed ref/id! $obj/$id!" unless $obj eq $obj_from_cache;
+            $UR::Context::all_objects_loaded->{$class}{$id} = $obj;
             print "KEEPING $obj.  Found $obj .\n";
             return;
         }
         else {
             if ($ENV{'UR_DEBUG_OBJECT_RELEASE'}) {
-                print STDERR "MEM DESTROY object $obj class ",$obj->class," id ",$obj->id,"\n";
+                print STDERR "MEM DESTROY object $obj class $class if $id\n";
```
Comment (tmooney) [inline lib/UR/Object.pm]: In the debug message `if` should stay `id`.

**Comment by brummett** [comment]
Fixed the typo

**Comment by tmooney** [comment]
This change generally seems good to me.  One strange thing I noticed when playing with it:

```perl
#!/usr/bin/perl
use strict; use warnings; use above 'UR'; use Genome;

sub foo {
    my $outer_variable;
    {
        my $guard = UR::Context::AutoUnloadPool->create();
        $outer_variable = Genome::Model::Build::ImportedReferenceSequence->get(name => 'GRCh37-lite-build37');
    }
    $outer_variable->run_by('x');
    return 1;
}

foo();
my $other_variable = Genome::Model::Build::ImportedReferenceSequence->get(name => 'GRCh37-lite-build37');
die unless $other_variable;
print STDERR $other_variable->run_by(), "\n";

exit;
```

In the old version, this would die in the assignment `run_by('x')` call, since `$outer_variable` becomes a `UR::DeletedRef`.  In this version, the update succeeds, but then we hit the `die` due to an undefined `$other_variable`.  (If there is no update to the object, then the subsequent `get()` succeeds as expected.)

**Comment by brummett** [comment]
The oddness wasn't with the AutoUnloadPool, but rather what happens with the changed objects after they go out of scope.  The original UR::Object::DESTROY() didn't have any checks for changed objects going out of scope.

These commits make the destructor save weakened/changed objects from disappearing from the object cache.  This destructor could use some refactoring help later; all these if/else branches is getting convoluted.

**Comment by brummett** [comment]
I think I've taken care of the strangeness you noticed.  I added your test and it's passing.

**Comment by tmooney** [comment]
+1

**Pull request merged**